platforms/chibios/drivers/vendor/RP/RP2040/ps2_vendor.c

```diff
@@ -0,0 +1,335 @@
+    sm_config_set_set_pins(&c, PS2_DATA_PIN, 2);
+    sm_config_set_out_pins(&c, PS2_DATA_PIN, 1);
+    sm_config_set_out_shift(&c, true, false, 10);
+    sm_config_set_in_shift(&c, true, true, 11);
+    sm_config_set_jmp_pin(&c, PS2_CLOCK_PIN);
+    sm_config_set_in_pins(&c, PS2_DATA_PIN);
+
+    // clang-format off
+    iomode_t pin_mode = PAL_RP_PAD_IE |
+                        PAL_RP_GPIO_OE |
+                        PAL_RP_PAD_SLEWFAST |
+                        PAL_RP_PAD_DRIVE12 |
+                        PAL_RP_IOCTRL_OEOVER_DRVPERI |
+                        (pio_idx == 0 ? PAL_MODE_ALTERNATE_PIO0 : PAL_MODE_ALTERNATE_PIO1);
+    // clang-format on
+
+    palSetLineMode(PS2_DATA_PIN, pin_mode);
+    palSetLineMode(PS2_CLOCK_PIN, pin_mode);
+
+#ifdef BUFFERED_MODE_ENABLE
+    pio_set_irq0_source_enabled(pio, pis_sm0_rx_fifo_not_empty + state_machine, true);
+#endif
+
+    pio_sm_init(pio, state_machine, offset, &c);
+
+#if defined(PS2_PIO_USE_PIO1)
+    nvicEnableVector(RP_PIO1_IRQ_0_NUMBER, CORTEX_MAX_KERNEL_PRIORITY);
+#else
+    nvicEnableVector(RP_PIO0_IRQ_0_NUMBER, CORTEX_MAX_KERNEL_PRIORITY);
+#endif
+
+    pio_sm_set_enabled(pio, state_machine, true);
+}
+
+static int bit_parity(int x) {
+    x ^= x >> 16;
+    x ^= x >> 8;
+    x ^= x >> 4;
+    x ^= x >> 2;
+    x ^= x >> 1;
+    return (~x) & 1;
+}
+
+uint8_t ps2_host_send(uint8_t data) {
+    uint32_t frame = 0b1000000000;
+    frame          = frame | data;
+
+    if (bit_parity(data)) {
+        frame = frame | (1 << 8);
+    }
+
+    pio_sm_put(pio, state_machine, frame);
+
+    msg_t msg = MSG_OK;
+    osalSysLock();
+    if (pio_sm_is_tx_fifo_full(pio, state_machine)) {
+        pio_set_irq0_source_enabled(pio, pis_sm0_tx_fifo_not_full + state_machine, true);
+        msg = osalThreadSuspendTimeoutS(&tx_thread, TIME_MS2I(100));
+        if (msg < MSG_OK) {
+            pio_set_irq0_source_enabled(pio, pis_sm0_tx_fifo_not_full + state_machine, false);
+            ps2_error = PS2_ERR_NODATA;
+            return 0;
+        }
+    }
+    osalSysUnlock();
+
+    return ps2_host_recv_response();
+}
+
+static uint8_t ps2_get_data_from_frame(uint32_t frame) {
+    uint8_t  data       = (frame >> 22) & 0xFF;
+    uint32_t start_bit  = (frame & 0b00000000001000000000000000000000) ? 1 : 0;
+    uint32_t parity_bit = (frame & 0b01000000000000000000000000000000) ? 1 : 0;
+    uint32_t stop_bit   = (frame & 0b10000000001000000000000000000000) ? 1 : 0;
+
+    if (start_bit != 0) {
+        ps2_error = PS2_ERR_STARTBIT1;
+        return 0;
+    }
+
+    if (parity_bit != bit_parity(data)) {
+        ps2_error = PS2_ERR_PARITY;
+        return 0;
+    }
+
+    if (stop_bit != 1) {
+        ps2_error = PS2_ERR_STARTBIT2;
+        return 0;
+    }
+
+    return data;
+}
+
+#ifndef BUFFERED_MODE_ENABLE
+uint8_t ps2_host_recv_response(void) {
+    msg_t msg = MSG_OK;
+    osalSysLock();
+    if (pio_sm_is_rx_fifo_empty(pio, state_machine)) {
```

Comment:
Same concern here, but you unlock properly here.

Reply:
Same question as upper, should it be blocking or not?

Reply:
See https://github.com/qmk/qmk_firmware/pull/17893#discussion_r937793225